Added EntityClimbEvent

zJanny opened the pull request:
This is the first time i have added an event to the API, feedback would be appreciated.

zJanny committed:
Added EntityClimbEvent

Owen1212055 commented on patches/api/0352-Added-EntityClimbEvent.patch:
```diff
@@ -0,0 +1,48 @@
++    public Block getClimbableBlock()
++    {
++        return climbableBlock;
++    }
```
Please stick to the same style, the opening bracket shouldn't be on a separate line.

Owen1212055 commented on patches/server/0842-Added-EntityClimbEvent.patch:
```diff
@@ -0,0 +1,63 @@
+ import org.bukkit.event.player.PlayerItemConsumeEvent;
+ // CraftBukkit end
+ 
+-import co.aikar.timings.MinecraftTimings; // Paper
```
I recommend undoing these import changes and simply using fully qualified imports. It cleans up the diff.

Owen1212055 commented on patches/server/0842-Added-EntityClimbEvent.patch:
```diff
@@ -0,0 +1,63 @@
+             if (d2 < 0.0D && !this.getFeetBlockState().is(Blocks.SCAFFOLDING) && this.isSuppressingSlidingDownLadder() && this instanceof net.minecraft.world.entity.player.Player) {
+                 d2 = 0.0D;
+             }
+-
```
Unneeded line change

Owen1212055 commented on patches/api/0352-Added-EntityClimbEvent.patch:
```diff
@@ -0,0 +1,48 @@
++import org.bukkit.event.entity.EntityEvent;
++import org.jetbrains.annotations.NotNull;
++
++public class EntityClimbEvent extends EntityEvent{
```
Additionally, you need to include javadocs. Look at other events for reference.

zJanny committed:
Update 0352-Added-EntityClimbEvent.patch

zJanny committed:
Update 0842-Added-EntityClimbEvent.patch

Machine-Maker commented on patches/server/0842-Added-EntityClimbEvent.patch:
```diff
@@ -0,0 +1,44 @@
++                org.bukkit.craftbukkit.block.CraftBlock craftBlock = org.bukkit.craftbukkit.block.CraftBlock.at(this.level, blockposition);
++                event = new io.papermc.paper.event.entity.EntityClimbEvent(this.getBukkitLivingEntity(), craftBlock);
++                event.callEvent();
++                //Paper end
```
If this is going to be a thing, the LivingEntity#onClimbable method is not the right place. CraftLivingEntity calls that when you call LivingEntity#isClimbing, and its called all over the place, in damage areas, collision checks, spigot's activation range. I feel like this event is going to fire seemingly randomly for plugins

zJanny replied:
It checks if the block the player is on has a climbable tag, so it shouldnt matter what calls the method. Or do i miss something?

Machine-Maker replied:
right, but lets say the event fires, and in the event, something calls LivingEntity#isClimbing, its going to fire again, and then you've got an infinite loop right there.

Or lets say within the span of 1 tick multiple places call that method, and since its the same tick, the player is on the climbable block, so the event has fired twice, or 4 times in the same tick.

Machine-Maker replied:
That method, I think, is supposed to a check *if* the entity is on a climbable thing. If this event is going to be implemented, it should be where some action is taken based on the return value of onClimbable being true

zJanny replied:
Okay thanks, i didnt think about that.

Machine-Maker reviewed (changes requested):
But you *can* get the block the player is climbing on. Look at what the onClimbable method does, it gets the block at the feed position and checks if its on the CLIMBABLE block tag.

zJanny committed:
Update 0352-Added-EntityClimbEvent.patch

zJanny committed:
Update 0842-Added-EntityClimbEvent.patch

This should now be in a function that may not cause a infinite loop. 
The event also now gets only triggered once per block, and spiders that just hang from the ceiling and dont move wont spam the event.
